### PR TITLE
Scaffold generate

### PR DESCRIPTION
**Pull request by fortmarek**
### Short description 📝

Follow-up PR to [scaffold load](https://github.com/tuist/tuist/pull/1092) where we added loading custom templates. This PR adds the final piece of a puzzle - and after it is merged, users will be finally able to define their own templates, which I am super excited about!

### Solution 📦

Added `TemplateGenerator` that handles generation logic of all files defined in `Template.swift`. Initially, it has contained a custom generation logic, but after some discussion we decided it is better to use already well-known templating system [stencil](https://github.com/stencilproject/Stencil)

### Implementation 👩‍💻👨‍💻

> Detail in a checklist the steps that you took to implement the PR.

- [x] Add `TemplateGenerator`
- [x] Add tests
- [x] Add fixture
- [x] Add documentation

### Notes ✏️

Unfortunately, we need to point to `master` branch for `Stencil` because they depend on `PathKit` 0.9.0 while `xcodeproj` depends on 1.0.0. Hopefully, a new release will be available soon but the repository does not seem to be highly maintained at the moment: https://github.com/stencilproject/Stencil/issues/285

### What's next 👀 
Next up is defining tuist built-in templates that can then be leveraged in `tuist init`.

### CODE REVIEW

File: Tests/TuistScaffoldTests/TemplateGeneratorTests.swift [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                             to: destinationPath,
+                             attributes: ["name": name])
+
+        // Then
+        try expectedFiles.forEach {
+            XCTAssertEqual(try FileHandler.shared.readTextFile($0.0), $0.1)
+        }
+    }
+
+    func test_file_rendered_wit_attributes() throws {
```
Comment (pepibumur) [inline Tests/TuistScaffoldTests/TemplateGeneratorTests.swift]:
```suggestion
func test_file_rendered_with_attributes() throws {
```

File: website/markdown/docs/usage/commands.mdx [hunk 2]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
 tuist edit --permanent
 ```
 
-That will generate a `Manfiest.xcodeproj` project that you can open manually.
+That will generate a `Manifest.xcodeproj` project that you can open manually.
+
+# Scaffold
+
+After project description helpers, `scaffold` is a next step to make it easier to create new modules in your app.
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]: I'd rephrase this a bit to not use "after xx" because that's a temporary reference that readers might not have context about. What about something along the lines of:

```
In projects with an established architecture, developers might want to bootstrap new components or features that are consistent with the project....
```

File: website/markdown/docs/usage/commands.mdx [hunk 3]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
...
+With `scaffold` you can define your own templates and soon also templates that Tuist provides to generate a basic structure of your module!
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]: I'd avoid using `soon` because it's also a temporary reference.

File: website/markdown/docs/usage/commands.mdx [hunk 4]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
...
+
+If you want to define your own template, we recommend you to run `tuist edit` - after that you can create a new directory at `Tuist/Templates/name_of_your_template` where `Tuist/Templates` is a directory where all your custom templates will live. Additionaly you have to create a `Template.swift` in that directory:
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]:
```suggestion
To define templates, you can run `tuist edit` and then create a directory under `Tuist/Templates` that represents your template. Templates need a manifest file, `Template.swift` that describes the template:
```

File: website/markdown/docs/usage/commands.mdx [hunk 5]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
...
+    ],
+)
+```
+
+This template would then be called like this:
+```bash
+tuist scaffold name_of_template --name Name --platform macos
+```
+
+You could also omit `--platform` argument since it is optional and the template provides a default value.
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]:
```
Since platform is an optional argument, we can also call the command without the `--platform macos` argument.
```

File: website/markdown/docs/usage/commands.mdx [hunk 6]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
...
+
+If you do not need any difficult logic, feel free to use `.string` for `files` parameter.
+Otherwise, you can leverage a powerful templating tool [Stencil](https://github.com/stencilproject/Stencil) via `.file` case.
+
+You can also leverage anything that you define in `ProjectDescriptionHelpers`, so you can share types and methods between your templates and manifests.
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]:
```suggestion
<Message info title="Project description helpers" description={`Templates can import [project description helpers](/docs/usage/helpers/). Just add `import ProjectDescriptionHelpers` at the top, and extract reusable logic into the helpers.`}/>
```

File: website/markdown/docs/usage/commands.mdx [hunk 7]
```diff
@@ -22,7 +22,47 @@ The project is deleted automatically once you are done with editing. If you wish
...
+```
+
+This template would then be called like this:
+```bash
+tuist scaffold name_of_template --name Name --platform macos
+```
+
+You could also omit `--platform` argument since it is optional and the template provides a default value.
+
+If you do not need any difficult logic, feel free to use `.string` for `files` parameter.
```
Comment (pepibumur) [inline website/markdown/docs/usage/commands.mdx]: I'd rephrase it a bit:
```
If `.string` and `.files` don't provide enough flexibility, you can leverage the [Stencil](https://github.com/stencilproject/Stencil) templating language via the `.file` case.
```

**Comment by fortmarek** [comment]
Thanks for the review @pepibumur! Stencil is being used [here](https://github.com/tuist/tuist/pull/1126/files#diff-fb02ff47a0650ef96689d0eb666afc5cR72) for the rendering of the templates. That is a good suggestion, maybe it'd be better to provide an example - maybe rather than forcing the folder with `.gitkeep` we could provide a simple example of a `Template` manifest?

**Comment by pepibumur** [comment]
> Thanks for the review @pepibumur! Stencil is being used here for the rendering of the templates.

👍 

> That is a good suggestion, maybe it'd be better to provide an example - so rather than forcing the folder with .gitkeep we could provide a simple example of a Template manifest?

Yeah, that's the other option that I was thinking about but I couldn't come up with a good example? A view model? An MVVM set of components?

**Comment by fortmarek** [comment]
I was thinking of a framework? It should be really dead simple - `Template.swift` that would generate a new `Project.swift` manifest and it'd take a name as an argument. You could then call `tuist scaffold framework --name MyFramework` and it'd create `MyFramework/Project.swift` file and directory.

**Comment by pepibumur** [comment]
> I was thinking of a framework? It should be really dead simple - Template.swift that would generate a new Project.swift manifest and it'd take a name as an argument. You could then call tuist scaffold framework --name MyFramework and it'd create MyFramework/Project.swift file and directory.

I love it ❤️

**Comment by fortmarek** [comment]
Added 👍 I have also had to make two changes:
1. stop verifying that the directory is empty - otherwise, you'd *always* have to specify `--path` which is not desirable
2. Added `.stencil` files to edit glob, so it appears in the editor project after running `tuist edit`

**Comment by fortmarek** [comment]
I am also not sure why the website build is failing, any hunch there @pepibumur ?

**Comment by pepibumur** [comment]
> I am also not sure why the website build is failing, any hunch there @pepibumur ?

I see that you already fixed it :). Feel free to merge the PR once CI is green 🚢

**Comment by AliSoftware** [comment]
@pepibumur @fortmarek Stencil 0.14.0 is now out and points to PathKit 1.0 🎉

**Comment by fortmarek** [comment]
Thanks! I'll make sure to update it, it's gonna make our lives more pleasant 🙇‍♂️
